In [1]:
#import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings("ignore")

In [2]:
#read the dataset
filename = pd.read_csv("StudentsPerformance.csv")
df = pd.DataFrame(filename)

In [3]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group C            some high school  free/reduced   
1  female        group B                 high school  free/reduced   
2  female        group B            some high school  free/reduced   
3  female        group B                some college      standard   
4  female        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none           0             17             10  
1                    none           8             24             23  
2                    none          18             32             28  
3                    none          11             38             32  
4                    none          22             39             33

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# 1. Find out how many males and females participated in the test?

In [5]:
df['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

# 2. What do you think about the students parental level of education?

In [6]:
df['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

Most of them are having some college level of education. The number of parents having some college level education and having associate's degree are almost equal. Only a few have master's degree.

# 3. Who scores the most on average for math, reading and writing based on

# ● Gender

In [7]:
df1=df.groupby('gender')['math score'].mean()
df1

gender
female    63.376448
male      68.821577
Name: math score, dtype: float64

On average for math, males scores the most than the females.

In [8]:
df2=df.groupby('gender')['reading score'].mean()
df2

gender
female    72.590734
male      65.545643
Name: reading score, dtype: float64

On average for reading, females scores the most than the males.

In [9]:
df3=df.groupby('gender')['writing score'].mean()
df3

gender
female    72.467181
male      63.446058
Name: writing score, dtype: float64

On average for writing, females scores the most than males.

# ● Test preparation course

In [10]:
df4=df.groupby('test preparation course')['math score'].mean()
df4

test preparation course
completed    69.96648
none         63.78972
Name: math score, dtype: float64

On average for math, those students whose test preparation course was completed scores the most.

In [11]:
df5=df.groupby('test preparation course')['reading score'].mean()
df5

test preparation course
completed    74.175978
none         66.417445
Name: reading score, dtype: float64

On average for reading, those students whose test preparation course was completed scores the most.

In [12]:
df6=df.groupby('test preparation course')['writing score'].mean()
df6

test preparation course
completed    74.684358
none         64.457944
Name: writing score, dtype: float64

On average for writing, those students whose test preparation course was completed scores the most.

# 4. What do you think about the scoring variation for math, reading and writing based on

# ● Gender

In [13]:
df1=df.groupby('gender')['math score'].std()
df1

gender
female    16.029928
male      14.556411
Name: math score, dtype: float64

In [14]:
df2=df.groupby('gender')['reading score'].std()
df2

gender
female    14.411018
male      14.149594
Name: reading score, dtype: float64

In [15]:
df3=df.groupby('gender')['writing score'].std()
df3

gender
female    14.844842
male      14.227225
Name: writing score, dtype: float64

Males are more consistent in scoring in math,reading and writing than females.

# ● Test preparation course

In [16]:
df4=df.groupby('test preparation course')['math score'].std()
df4

test preparation course
completed    14.521847
none         15.705689
Name: math score, dtype: float64

In [17]:
df5=df.groupby('test preparation course')['reading score'].std()
df5

test preparation course
completed    13.537572
none         14.608896
Name: reading score, dtype: float64

In [18]:
df6=df.groupby('test preparation course')['writing score'].std()
df6

test preparation course
completed    13.236412
none         15.041667
Name: writing score, dtype: float64

The students whose test preparation course was completed scored in all fields more consistently.

# 5. The management needs your help to give bonus points to the top 25% of students based on their maths score, so how will you help the management to achieve this?

In [19]:
df['math score'].value_counts().sum()

1000

In [20]:
1000*25/100

250.0

Therefore the top 250 students comes under top 25%

In [21]:
top25=df.sort_values(by=['math score'],ascending=False).head(250)
top25

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
856    male        group E            some high school      standard   
855    male        group E          associate's degree  free/reduced   
854    male        group D            some high school      standard   
853    male        group D          associate's degree  free/reduced   
852    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
856               completed          77             76             77  
855               completed          77             69             68  
854               completed          77             68             69  
853                    none          77             78             73  
852                    none          77             62             64  

[250 rows x 8 columns]

Here we could sort the values by maths score in descending order. On this sorted data, the first 250 students comes under top25%.

# Or based on Math score

In [23]:
top_df_math=df['math score'].quantile([0.75])

In [24]:
top_df_math

0.75    77.0
Name: math score, dtype: float64

So students who scored more than 77 in the maths score comes under top 25%

In [36]:
top_df_math25=df[df['math score']>77]

In [37]:
top_df_math25

gender race/ethnicity parental level of education         lunch  \
430  female        group A                some college      standard   
431  female        group B           bachelor's degree  free/reduced   
432  female        group D           bachelor's degree      standard   
433  female        group D                 high school      standard   
434  female        group D           bachelor's degree  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
430               completed          78             87             91  
431                    none          78             79             76  
432                    none          78             82             79  
433                    none          78             81             80  
434                    none          78             90             93  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[231 rows x 8 columns]

There are 231 students who comes under top 25% based on their maths score.